File: lib/features/landlord/controllers/post_controller.dart
import 'dart:async';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:flutter/widgets.dart';
import 'package:get/get.dart';
import '../../../core/network/api_checker.dart';
import '../../../core/services/notification_service.dart';
import '../../../core/utils/app_logger.dart';
import '../../../core/utils/imgbb_service.dart';
import '../../auth/controllers/auth_controller.dart';
import '../models/post_model.dart';
import '../repositories/post_repo.dart';
import 'package:mess_finder/features/notifications/models/app_notification_model.dart';
import '../../../core/services/location_service.dart';
import '../../../core/services/api_service.dart';
import 'package:geolocator/geolocator.dart';

class PostController extends GetxController {
  final PostRepository _postRepo = PostRepository();

  // Location
  final Rx<Position?> userLocation = Rx<Position?>(null);

  // Reactive Lists
  final RxList<PostModel> allPosts = <PostModel>[].obs;
  final RxList<PostModel> myPosts = <PostModel>[].obs;
  final RxSet<String> savedPostIds = <String>{}.obs;
  final RxBool isLoading = true.obs;
  final RxBool isFetchingMore = false.obs;
  final RxBool hasMorePosts = true.obs;

  String? lastDocument;
  int currentPage = 1;
  final int postLimit = 10;
  final ScrollController feedScrollController = ScrollController();

  // Cache for Landlord Profiles to optimize scrolling
  final Map<String, Map<String, dynamic>> landlordProfilesCache = {};

  // Future cache to prevent concurrent identical requests
  final Map<String, Future<Map<String, dynamic>?>> _profileFetchFutures = {};

  Future<Map<String, dynamic>?> getLandlordProfile(String uid) async {
    // If it's the current user, always return the latest local data
    if (Get.isRegistered<AuthController>()) {
      final auth = Get.find<AuthController>();
      final user = auth.currentUser.value;
      if (user != null && user.uid == uid) {
        return user.toMap();
      }
    }

    if (landlordProfilesCache.containsKey(uid)) {
      return landlordProfilesCache[uid];
    }

    // If a fetch is already in progress for this uid, await and return it
    if (_profileFetchFutures.containsKey(uid)) {
      return await _profileFetchFutures[uid];
    }

    // Otherwise, create a new fetch Future and store it
    final fetchFuture = _fetchProfileFromFirestore(uid);
    _profileFetchFutures[uid] = fetchFuture;

    final result = await fetchFuture;

    // Once done, remove from pending futures
    _profileFetchFutures.remove(uid);

    return result;
  }

  Future<Map<String, dynamic>?> _fetchProfileFromFirestore(String uid) async {
    try {
      final res = await ApiService().dio.get('/auth/user/$uid');
      if (res.statusCode == 200 && res.data != null) {
        final data = Map<String, dynamic>.from(res.data);
        final photo = data['profile_image'] ?? data['photoUrl'];
        final map = {
          'uid': data['uid'],
          'name': data['name'] ?? 'Landlord',
          'phone': data['phone'] ?? '',
          'photoUrl': photo,
          'profile_image': photo,
          'isPaid': data['status'] == 'active' || data['isPaid'] == true,
          'role': data['role'] ?? 'landlord',
        };
        landlordProfilesCache[uid] = map;
        return map;
      }
    } catch (e) {
      debugPrint('Error fetching landlord profile: $e');
    }
    return {'uid': uid, 'name': 'Landlord', 'phone': ''};
  }

  // Filter states for Bachelor Feed
  final RxString selectedGenderFilter =
      'all'.obs; // 'all', 'male', 'female', 'both'
  final RxInt selectedBudgetFilter = 0.obs; // 0=All, 4000, 6000, 8000
  final RxString searchQuery = ''.obs;
  Timer? _searchDebounce;

  void updateSearchQuery(String query) {
    if (_searchDebounce?.isActive ?? false) _searchDebounce!.cancel();
    _searchDebounce = Timer(const Duration(milliseconds: 500), () {
      searchQuery.value = query;
    });
  }

  final RxString mapSearchQuery = ''.obs;
  Timer? _mapSearchDebounce;

  void updateMapSearchQuery(String query) {
    if (_mapSearchDebounce?.isActive ?? false) _mapSearchDebounce!.cancel();
    _mapSearchDebounce = Timer(const Duration(milliseconds: 500), () {
      mapSearchQuery.value = query;
    });
  }

  final RxString selectedDivisionFilter = 'All'.obs;
  final RxString selectedDistrictFilter = 'All'.obs;

  bool isSaved(String postId) => savedPostIds.contains(postId);

  void toggleSavePost(String postId) {
    if (savedPostIds.contains(postId)) {
      savedPostIds.remove(postId);
      _updateSavedPostsList();
      _syncSavedPostsToLocal();
      Get.snackbar(
        'Favorites',
        'Removed from your favorites list',
        snackPosition: SnackPosition.BOTTOM,
        duration: const Duration(seconds: 2),
      );
    } else {
      savedPostIds.add(postId);
      _updateSavedPostsList();
      _syncSavedPostsToLocal();
      Get.snackbar(
        'Favorites ❤️',
        'Added to your favorites list! You can find it in the Favorites tab.',
        snackPosition: SnackPosition.BOTTOM,
        duration: const Duration(seconds: 2),
      );
    }
  }

  final RxList<PostModel> savedPosts = <PostModel>[].obs;

  void _updateSavedPostsList() {
    savedPosts.assignAll(
      allPosts.where((post) => savedPostIds.contains(post.postId)).toList(),
    );
    AppLogger.i(
      'DEBUG: _updateSavedPostsList called. allPosts=${allPosts.length}, savedIds=${savedPostIds.length}, result=${savedPosts.length}',
    );
  }

  @override
  void onInit() {
    super.onInit();
    _initPosts();

    feedScrollController.addListener(() {
      if (feedScrollController.position.pixels >=
          feedScrollController.position.maxScrollExtent - 200) {
        if (!isLoading.value && !isFetchingMore.value) {
          loadMorePosts();
        }
      }
    });
  }

  Future<void> fetchUserLocation() async {
    final position = await LocationService.getCurrentLocation();
    if (position != null) {
      userLocation.value = position;
    }
  }

  Future<void> _initPosts() async {
    isLoading.value = true;
    try {
      // Demo post deletion removed so they don't disappear on reload
    } catch (e) {
      // Ignore error
    }

    // Attempt to fetch user location silently in the background
    fetchUserLocation();

    _loadSavedPostsFromLocal();

    await fetchInitialPosts();

    // Listen to current user's posts
    if (Get.isRegistered<AuthController>()) {
      final auth = Get.find<AuthController>();
      ever(auth.currentUser, (user) {
        _loadSavedPostsFromLocal();
        if (user != null) {
          fetchMyPosts();
        } else {
          myPosts.clear();
        }
      });
      if (auth.currentUser.value != null) {
        fetchMyPosts();
      }
    }
  }

  Future<void> fetchMyPosts() async {
    try {
      if (Get.isRegistered<AuthController>()) {
        final auth = Get.find<AuthController>();
        if (auth.currentUser.value == null) {
          myPosts.clear();
          return;
        }
      }
      final posts = await _postRepo.getLandlordPosts();
      myPosts.assignAll(posts);
    } catch (e) {
      debugPrint('Error in fetchMyPosts: $e');
    }
  }

  Future<void> refreshPosts() async {
    await Future.wait([
      fetchInitialPosts(),
      fetchMyPosts(),
    ]);
  }

  Future<void> fetchInitialPosts() async {
    isLoading.value = true;
    hasMorePosts.value = true;
    currentPage = 1;
    lastDocument = null;
    final result = await _postRepo.getPaginatedPosts(
      limit: postLimit,
      page: 1,
      division: selectedDivisionFilter.value == 'All'
          ? null
          : selectedDivisionFilter.value,
      district: selectedDistrictFilter.value == 'All'
          ? null
          : selectedDistrictFilter.value,
    );
    final List<PostModel> fetchedPosts = result['posts'] as List<PostModel>;
    lastDocument = result['lastDocument'] as String?;

    if (fetchedPosts.length < postLimit) {
      hasMorePosts.value = false;
    }

    allPosts.assignAll(fetchedPosts);
    _updateSavedPostsList();
    isLoading.value = false;
  }

  void updateSearch(String query) {
    searchQuery.value = query;
  }

  void updateLocationFilter(String division, String district) {
    selectedDivisionFilter.value = division;
    selectedDistrictFilter.value = district;
    fetchInitialPosts();
  }

  Future<void> loadMorePosts() async {
    if (isFetchingMore.value || !hasMorePosts.value) return;
    isFetchingMore.value = true;
    final nextPage = currentPage + 1;
    final result = await _postRepo.getPaginatedPosts(
      limit: postLimit,
      page: nextPage,
      startAfter: lastDocument,
      division: selectedDivisionFilter.value == 'All'
          ? null
          : selectedDivisionFilter.value,
      district: selectedDistrictFilter.value == 'All'
          ? null
          : selectedDistrictFilter.value,
    );

    final List<PostModel> newPosts = result['posts'] as List<PostModel>;
    lastDocument = result['lastDocument'] as String?;

    if (newPosts.length < postLimit) {
      hasMorePosts.value = false;
    }

    if (newPosts.isNotEmpty) {
      currentPage = nextPage;
      allPosts.addAll(newPosts);
      _updateSavedPostsList();
    }

    isFetchingMore.value = false;
  }

  @override
  void onClose() {
    _searchDebounce?.cancel();
    super.onClose();
  }

  Future<void> _loadSavedPostsFromLocal() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final savedList = prefs.getStringList('savedPosts');
      if (savedList != null) {
        savedPostIds.assignAll(savedList);
        _updateSavedPostsList();
      }
    } catch (e) {
      AppLogger.e('Failed to load favorite posts from SharedPreferences: $e', e, null, 'POST_CTRL');
    }
  }

  Future<void> _syncSavedPostsToLocal() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setStringList('savedPosts', savedPostIds.toList());
    } catch (e) {
      AppLogger.e('Failed to save favorite post to SharedPreferences: $e', e, null, 'POST_CTRL');
    }
  }

  // Computed list for Bachelor Feed after applying Search & Filters
  List<PostModel> get filteredPosts {
    final filtered = allPosts.where((post) {
      // 0. Auto-expiry filter (15 days)
      if (post.createdAt != null) {
        final age = DateTime.now().difference(post.createdAt!).inDays;
        if (age > 15) {
          return false;
        }
      }

      // 1. Search Query filter (title or address)
      if (searchQuery.value.isNotEmpty) {
        final query = searchQuery.value.toLowerCase();
        final matchTitle = post.title.toLowerCase().contains(query);
        final matchAddress = post.address.toLowerCase().contains(query);
        final matchDistrict = post.district.toLowerCase().contains(query);
        final matchDivision = post.division.toLowerCase().contains(query);
        if (!matchTitle && !matchAddress && !matchDistrict && !matchDivision) {
          return false;
        }
      }
      // 2. Gender filter
      if (selectedGenderFilter.value != 'all') {
        final postType = post.bachelorType.trim().toLowerCase();
        final filterType = selectedGenderFilter.value.trim().toLowerCase();
        if (postType != filterType && postType != 'both') {
          return false;
        }
      }
      // 3. Budget filter
      if (selectedBudgetFilter.value > 0) {
        if (post.rent > selectedBudgetFilter.value) {
          return false;
        }
      }
      return true;
    }).toList();

    // 4. Always sort by newest first (Recent posts at the top)
    filtered.sort((a, b) {
      if (a.createdAt == null || b.createdAt == null) return 0;
      return b.createdAt!.compareTo(a.createdAt!);
    });

    return filtered;
  }

  List<PostModel> get mapFilteredPosts {
    List<PostModel> filtered = List.from(allPosts);

    // Apply same filters (budget, gender, location) BUT use mapSearchQuery instead of searchQuery

    // 1. Map Search Query
    if (mapSearchQuery.value.trim().isNotEmpty) {
      final query = mapSearchQuery.value.trim().toLowerCase();
      filtered = filtered.where((post) {
        return post.title.toLowerCase().contains(query) ||
            post.address.toLowerCase().contains(query) ||
            post.division.toLowerCase().contains(query) ||
            post.district.toLowerCase().contains(query);
      }).toList();
    }

    // 2. Gender
    if (selectedGenderFilter.value != 'all') {
      filtered = filtered.where((post) {
        final bType = post.bachelorType.toLowerCase();
        if (selectedGenderFilter.value == 'male') {
          return bType.contains('male') || bType.contains('boy');
        } else if (selectedGenderFilter.value == 'female') {
          return bType.contains('female') || bType.contains('girl');
        }
        return true;
      }).toList();
    }

    // 3. Budget
    if (selectedBudgetFilter.value > 0) {
      filtered = filtered.where((post) {
        return post.rent <= selectedBudgetFilter.value;
      }).toList();
    }

    return filtered;
  }

  // Add a new Mess Post (Landlord)
  Future<bool> addMessPost({
    required String title,
    required double rent,
    required String address,
    required int seatCount,
    String? seatDescription,
    required String division,
    required String district,
    String? ownerPhone,
    required String bachelorType,
    String preferredTenant = 'Student / Job holder',
    required List<String> facilities,
    required List<String> images,
    String? videoPath,
    String? trxId,
    String? senderNumber,
    double latitude = 23.8103,
    double longitude = 90.4125,
  }) async {
    final auth = Get.find<AuthController>();
    final user = auth.currentUser.value;
    if (user == null) {
      ApiChecker.showError('Please log in again');
      return false;
    }

    try {
      isLoading.value = true;

      final storageService = ImgbbService();
      final List<String> finalImageUrls = [];

      for (String path in images) {
        if (path.startsWith('http://') || path.startsWith('https://')) {
          finalImageUrls.add(path);
        } else {
          final url = await storageService.uploadImage(path);
          if (url != null) finalImageUrls.add(url);
        }
      }

      if (finalImageUrls.isEmpty && images.isNotEmpty) {
        throw 'Image upload failed! Please check your internet connection and try again.\n\nTip: Cloudinary preset "messfinder_preset" must be set to "Unsigned" in Cloudinary dashboard.';
      }

      String? uploadedVideoUrl;
      if (videoPath != null && videoPath.isNotEmpty) {
        if (videoPath.startsWith('http://') || videoPath.startsWith('https://')) {
          uploadedVideoUrl = videoPath;
        } else {
          uploadedVideoUrl = await storageService.uploadVideo(videoPath);
        }
      }

      final newPost = PostModel(
        postId: '',
        ownerUid: user.uid,
        ownerPhone: (ownerPhone != null && ownerPhone.trim().isNotEmpty)
            ? ownerPhone.trim()
            : user.phone,
        title: title,
        rent: rent,
        address: address,
        latitude: latitude,
        longitude: longitude,
        images: finalImageUrls,
        videoUrl: uploadedVideoUrl,
        seatCount: seatCount,
        seatDescription: seatDescription,
        division: division,
        district: district,
        bachelorType: bachelorType,
        preferredTenant: preferredTenant,
        facilities: facilities,
        isAvailable: true,
        isPublished: true,
        paymentStatus: 'approved',
        paymentTrxId: trxId,
        senderNumber: senderNumber,
        createdAt: DateTime.now(),
      );

      await _postRepo.addPost(newPost);

      // Broadcast push notification to bachelors (and others) about the new room
      try {
        final loc = address.split(',').first;
        final titleStr = 'New Room Available! 🏠';
        final bodyStr =
            'A new $bachelorType room is available in $loc. Rent: $rent';

        NotificationService().storeNotification(
          AppNotificationModel(
            id: '',
            title: titleStr,
            body: bodyStr,
            type: NotificationType.newPost,
            receiverUid: 'all',
            senderUid: user.uid,
            createdAt: DateTime.now(),
          ),
        );

        NotificationService().sendPushToTopic(
          topic: 'all_users',
          title: titleStr,
          body: bodyStr,
          data: {'type': 'new_post'},
        );
      } catch (e) {
        // ignore notification errors
      }

      await Future.wait([
        fetchInitialPosts(),
        fetchMyPosts(),
      ]);

      return true;
    } catch (e) {
      ApiChecker.showError(e.toString());
      return false;
    } finally {
      isLoading.value = false;
    }
  }

  // Update an existing Mess Post
  Future<bool> updateMessPost(PostModel updatedPost, {String? newVideoPath}) async {
    try {
      isLoading.value = true;

      final storageService = ImgbbService();
      final List<String> finalImageUrls = [];
      bool anyLocalUploadFailed = false;

      for (String path in updatedPost.images) {
        if (path.startsWith('http://') || path.startsWith('https://')) {
          finalImageUrls.add(path);
        } else {
          final url = await storageService.uploadImage(path);
          if (url != null) {
            finalImageUrls.add(url);
          } else {
            anyLocalUploadFailed = true;
          }
        }
      }

      // Only fail if there are no images at all after processing
      if (finalImageUrls.isEmpty) {
        throw 'Failed to upload images. Please check your internet connection.';
      }

      if (anyLocalUploadFailed) {
        ApiChecker.showError('Some images failed to upload and were skipped.');
      }

      String? finalVideoUrl = updatedPost.videoUrl;
      if (newVideoPath != null && newVideoPath.isNotEmpty) {
        if (newVideoPath.startsWith('http://') || newVideoPath.startsWith('https://')) {
          finalVideoUrl = newVideoPath;
        } else {
          final uploadedUrl = await storageService.uploadVideo(newVideoPath);
          if (uploadedUrl != null) {
            finalVideoUrl = uploadedUrl;
          }
        }
      }

      final finalPost = updatedPost.copyWith(images: finalImageUrls, videoUrl: finalVideoUrl);

      // Instant optimistic UI update
      final myIdx = myPosts.indexWhere((p) => p.postId == finalPost.postId);
      if (myIdx != -1) {
        myPosts[myIdx] = finalPost;
        myPosts.refresh();
      }
      final allIdx = allPosts.indexWhere((p) => p.postId == finalPost.postId);
      if (allIdx != -1) {
        allPosts[allIdx] = finalPost;
        allPosts.refresh();
      }

      await _postRepo.updatePost(finalPost);
      fetchMyPosts();
      return true;
    } catch (e) {
      ApiChecker.showError(e.toString());
      return false;
    } finally {
      isLoading.value = false;
    }
  }

  // Delete Mess Post
  Future<void> deleteMessPost(String postId) async {
    // Instant optimistic removal from UI
    myPosts.removeWhere((p) => p.postId == postId);
    allPosts.removeWhere((p) => p.postId == postId);
    savedPostIds.remove(postId);
    _updateSavedPostsList();
    try {
      await _postRepo.deletePost(postId);
      ApiChecker.showSuccess('Room listing deleted successfully');
    } catch (e) {
      fetchMyPosts();
      ApiChecker.showError(e.toString());
    }
  }

  // Toggle availability
  Future<void> togglePostStatus(String postId, bool currentStatus) async {
    final newStatus = !currentStatus;
    // Instant optimistic update in memory
    final myIdx = myPosts.indexWhere((p) => p.postId == postId);
    if (myIdx != -1) {
      myPosts[myIdx] = myPosts[myIdx].copyWith(isAvailable: newStatus);
      myPosts.refresh();
    }
    final allIdx = allPosts.indexWhere((p) => p.postId == postId);
    if (allIdx != -1) {
      allPosts[allIdx] = allPosts[allIdx].copyWith(isAvailable: newStatus);
      allPosts.refresh();
    }
    try {
      await _postRepo.toggleAvailability(postId, currentStatus);
      ApiChecker.showSuccess(
        currentStatus ? 'Listing deactivated' : 'Listing activated',
      );
    } catch (e) {
      fetchMyPosts();
      ApiChecker.showError(e.toString());
    }
  }
}

File: backend/controllers/postController.js
const pool = require('../config/db');

// Add new post
exports.createPost = async (req, res) => {
  const {
    title, rent, address, latitude, longitude, images, videoUrl,
    seatCount, seatDescription, division, district, bachelorType,
    preferredTenant, facilities, ownerPhone, paymentTrxId, senderNumber,
    isAvailable, isPublished, paymentStatus
  } = req.body;

  try {
    const newPost = await pool.query(
      `INSERT INTO posts (
        owner_uid, title, rent, address, latitude, longitude, images, video_url,
        seat_count, seat_description, division, district, bachelor_type,
        preferred_tenant, facilities, owner_phone, payment_trx_id, sender_number,
        is_available, is_published, payment_status
      ) VALUES (
        $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21
      ) RETURNING *`,
      [
        req.user.uid,
        title,
        rent,
        address,
        latitude || 23.8103,
        longitude || 90.4125,
        JSON.stringify(images || []),
        videoUrl || null,
        seatCount || 1,
        seatDescription || null,
        division || 'Dhaka',
        district || 'Dhaka',
        bachelorType || 'male',
        preferredTenant || 'Student / Job holder',
        JSON.stringify(facilities || []),
        ownerPhone || null,
        paymentTrxId || null,
        senderNumber || null,
        isAvailable !== undefined ? isAvailable : true,
        isPublished !== undefined ? isPublished : true,
        paymentStatus || 'approved'
      ]
    );

    const post = newPost.rows[0];
    post.images = typeof post.images === 'string' ? JSON.parse(post.images) : post.images;
    post.facilities = typeof post.facilities === 'string' ? JSON.parse(post.facilities) : post.facilities;

    res.status(201).json(post);
  } catch (error) {
    console.error('Error creating post:', error);
    res.status(500).json({ error: 'Server error creating post' });
  }
};

// Get all published posts (with optional filtering)
exports.getPosts = async (req, res) => {
  try {
    const { district, division, bachelorType, ownerUid, owner_uid } = req.query;
    const targetOwner = ownerUid || owner_uid;
    
    let query = 'SELECT * FROM posts WHERE is_published = true AND is_available = true';
    const params = [];
    let paramIndex = 1;

    if (targetOwner) {
      query += ` AND owner_uid = $${paramIndex++}`;
      params.push(targetOwner);
    }
    if (district && district !== 'All') {
      query += ` AND district = $${paramIndex++}`;
      params.push(district);
    }
    if (division && division !== 'All') {
      query += ` AND division = $${paramIndex++}`;
      params.push(division);
    }
    if (bachelorType && bachelorType !== 'all') {
      query += ` AND bachelor_type = $${paramIndex++}`;
      params.push(bachelorType);
    }

    query += ' ORDER BY created_at DESC';

    const result = await pool.query(query, params);
    
    const posts = result.rows.map(row => ({
      ...row,
      images: typeof row.images === 'string' ? JSON.parse(row.images) : row.images,
      facilities: typeof row.facilities === 'string' ? JSON.parse(row.facilities) : row.facilities
    }));

    res.status(200).json(posts);
  } catch (error) {
    console.error('Error fetching posts:', error);
    res.status(500).json({ error: 'Server error fetching posts' });
  }
};

// Get single post
exports.getPostById = async (req, res) => {
  try {
    const result = await pool.query('SELECT * FROM posts WHERE post_id = $1', [req.params.id]);
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Post not found' });
    }
    const row = result.rows[0];
    row.images = typeof row.images === 'string' ? JSON.parse(row.images) : row.images;
    row.facilities = typeof row.facilities === 'string' ? JSON.parse(row.facilities) : row.facilities;
    
    res.status(200).json(row);
  } catch (error) {
    console.error('Error fetching post by ID:', error);
    res.status(500).json({ error: 'Server error' });
  }
};

// Get My Posts (for logged-in landlord)
exports.getMyPosts = async (req, res) => {
  try {
    const result = await pool.query('SELECT * FROM posts WHERE owner_uid = $1 ORDER BY created_at DESC', [req.user.uid]);
    const posts = result.rows.map(row => ({
      ...row,
      images: typeof row.images === 'string' ? JSON.parse(row.images) : row.images,
      facilities: typeof row.facilities === 'string' ? JSON.parse(row.facilities) : row.facilities
    }));
    res.status(200).json(posts);
  } catch (error) {
    console.error('Error fetching my posts:', error);
    res.status(500).json({ error: 'Server error' });
  }
};

// Update Post (supports full edit, status toggle, admin approve/reject)
exports.updatePost = async (req, res) => {
  const { id } = req.params;
  const {
    title, rent, address, latitude, longitude, images, videoUrl,
    seatCount, seatDescription, division, district, bachelorType,
    preferredTenant, facilities, ownerPhone, isAvailable, isPublished,
    paymentStatus, paymentTrxId, senderNumber
  } = req.body;

  try {
    // Check if post exists and user has rights (owner or admin)
    const existingPost = await pool.query('SELECT * FROM posts WHERE post_id = $1', [id]);
    if (existingPost.rows.length === 0) {
      return res.status(404).json({ error: 'Post not found' });
    }

    const userCheck = await pool.query('SELECT role FROM users WHERE uid = $1', [req.user.uid]);
    const userRole = userCheck.rows[0]?.role;
    const isOwner = existingPost.rows[0].owner_uid === req.user.uid;
    const isAdmin = userRole === 'admin';

    if (!isOwner && !isAdmin) {
      return res.status(403).json({ error: 'Not authorized to update this post' });
    }

    const current = existingPost.rows[0];

    const updatedImages = images !== undefined ? JSON.stringify(images) : current.images;
    const updatedFacilities = facilities !== undefined ? JSON.stringify(facilities) : current.facilities;

    const result = await pool.query(
      `UPDATE posts SET 
        title = COALESCE($1, title),
        rent = COALESCE($2, rent),
        address = COALESCE($3, address),
        latitude = COALESCE($4, latitude),
        longitude = COALESCE($5, longitude),
        images = COALESCE($6::jsonb, images),
        video_url = COALESCE($7, video_url),
        seat_count = COALESCE($8, seat_count),
        seat_description = COALESCE($9, seat_description),
        division = COALESCE($10, division),
        district = COALESCE($11, district),
        bachelor_type = COALESCE($12, bachelor_type),
        preferred_tenant = COALESCE($13, preferred_tenant),
        facilities = COALESCE($14::jsonb, facilities),
        owner_phone = COALESCE($15, owner_phone),
        is_available = COALESCE($16, is_available),
        is_published = COALESCE($17, is_published),
        payment_status = COALESCE($18, payment_status),
        payment_trx_id = COALESCE($19, payment_trx_id),
        sender_number = COALESCE($20, sender_number),
        updated_at = CURRENT_TIMESTAMP
       WHERE post_id = $21 RETURNING *`,
      [
        title,
        rent,
        address,
        latitude,
        longitude,
        updatedImages,
        videoUrl,
        seatCount,
        seatDescription,
        division,
        district,
        bachelorType,
        preferredTenant,
        updatedFacilities,
        ownerPhone,
        isAvailable,
        isPublished,
        paymentStatus,
        paymentTrxId,
        senderNumber,
        id
      ]
    );

    const post = result.rows[0];
    post.images = typeof post.images === 'string' ? JSON.parse(post.images) : post.images;
    post.facilities = typeof post.facilities === 'string' ? JSON.parse(post.facilities) : post.facilities;

    res.status(200).json(post);
  } catch (error) {
    console.error('Error updating post:', error);
    res.status(500).json({ error: 'Server error updating post' });
  }
};

// Delete Post
exports.deletePost = async (req, res) => {
  const { id } = req.params;
  try {
    const existingPost = await pool.query('SELECT * FROM posts WHERE post_id = $1', [id]);
    if (existingPost.rows.length === 0) {
      return res.status(404).json({ error: 'Post not found' });
    }

    const userCheck = await pool.query('SELECT role FROM users WHERE uid = $1', [req.user.uid]);
    const userRole = userCheck.rows[0]?.role;
    const isOwner = existingPost.rows[0].owner_uid === req.user.uid;
    const isAdmin = userRole === 'admin';

    if (!isOwner && !isAdmin) {
      return res.status(403).json({ error: 'Not authorized to delete this post' });
    }

    await pool.query('DELETE FROM posts WHERE post_id = $1', [id]);
    res.status(200).json({ message: 'Deleted successfully' });
  } catch (error) {
    console.error('Error deleting post:', error);
    res.status(500).json({ error: 'Server error' });
  }
};

// Toggle Availability
exports.toggleAvailability = async (req, res) => {
  const { id } = req.params;
  const { isAvailable } = req.body;
  try {
    const result = await pool.query(
      'UPDATE posts SET is_available = $1, updated_at = CURRENT_TIMESTAMP WHERE post_id = $2 AND owner_uid = $3 RETURNING *',
      [isAvailable, id, req.user.uid]
    );
    if (result.rows.length === 0) return res.status(403).json({ error: 'Not authorized or post not found' });
    
    const post = result.rows[0];
    post.images = typeof post.images === 'string' ? JSON.parse(post.images) : post.images;
    post.facilities = typeof post.facilities === 'string' ? JSON.parse(post.facilities) : post.facilities;

    res.status(200).json(post);
  } catch (error) {
    console.error('Error toggling post availability:', error);
    res.status(500).json({ error: 'Server error' });
  }
};
